### PR TITLE
Only toggle aria-hidden for DOM nodes coming from grommet

### PR DESCRIPTION
**Pull request by taysea**
#### What does this PR do?
This PR updates how Grommet toggles `aria-hidden` values on DOM elements. Now, Grommet will only touch elements that have been added to the DOM by grommet. It will not touch other children of `<body>` that may not be related to Grommet.

This is done by creating a context called `RootsContext` that holds an array of refs of any Grommet related elements appended as a child of `body`. By default, it holds the ref of the `<Grommet>` tag, but any time a FocusedContainer is added (Drop or Modal Layer), the ref of that div is added to the array. Instead of iterating on all body children, we now iterate over this Grommet specific array.

Adjustments were also made to the useEffects in `FocusedContainer` to separate unrelated functionality.

Additionally, the `removeTrap` function was enhanced to fix a bug where all elements with`aria-hidden="true"` were toggling back to `aria-hidden="false"`.

#### Where should the reviewer start?
src/js/components/Grommet/Grommet.js

#### What testing has been done on this PR?
Tested locally in storybook and jest.

I've confirmed that in the case that you have nested Grommets, the inner Grommet will not toggle `aria-hidden` values for an outer Grommet.

#### How should this be manually tested?
Interact with any Drop story and open dev tools. `aria-hidden` should not be added to body children not related to Grommet.

#### What are the relevant issues?
Closes https://github.com/grommet/grommet/issues/4739
 
#### Do the grommet docs need to be updated?
No.

#### Should this PR be mentioned in the release notes?
No.

#### Is this change backwards compatible or is it a breaking change?
Backwards compatible.

### CODE REVIEW

**Comment by ericsoderberghp** [comment]
> One edge case to account for is if the user does not wrap their application in a `<Grommet>` (as was the case in some jest tests). In this case, we add a `data-g-id="grommet"` to the FocusedContainer which `getBodyChildElements` will look for. Since the GrommetContext is not available, we need some way to mark the element directly in the DOM.

I don't think we should be handling this case. I think a `<Grommet />` should be required for us to manage `aria-hidden` elsewhere. We will need this behavior to handle multiple versions of grommet in the same browser DOM.

**Comment by taysea** [comment]
> > One edge case to account for is if the user does not wrap their application in a `<Grommet>` (as was the case in some jest tests). In this case, we add a `data-g-id="grommet"` to the FocusedContainer which `getBodyChildElements` will look for. Since the GrommetContext is not available, we need some way to mark the element directly in the DOM.
> 
> I don't think we should be handling this case. I think a `<Grommet />` should be required for us to manage `aria-hidden` elsewhere. We will need this behavior to handle multiple versions of grommet in the same browser DOM.

Okay that's fair. I will remove this functionality then.

File: src/js/components/Form/__tests__/__snapshots__/Form-test-uncontrolled.js.snap [hunk 1]
```diff
@@ -1941,6 +1941,7 @@ exports[`Form uncontrolled reset clears select, checkbox, radiobuttongroup 1`] =
 }
 
 <div
+  aria-hidden="false"
```
Comment (ericsoderberghp) [inline src/js/components/Form/__tests__/__snapshots__/Form-test-uncontrolled.js.snap:1944]: Why are we adding this? It seems like it isn't needed.

File: src/js/components/Grommet/Grommet.js [hunk 2]
```diff
@@ -93,15 +95,23 @@ const Grommet = forwardRef((props, ref) => {
     deviceResponsive(userAgent, theme) ||
     theme.global.deviceBreakpoints.tablet;
 
+  const grommetRef = useForwardedRef(ref);
+  // When toggling aria-hidden values, we only want to affect elements
+  // in the DOM that come from Grommet, so we track those elemnts in this
+  // context value. See FocusedContainer.js
+  const bodyChildrenFromGrommet = [grommetRef.current];
```
Comment (ericsoderberghp) [inline src/js/components/Grommet/Grommet.js:102]: What if I have nested `<Grommet />` elements?

File: src/js/components/Grommet/GrommetContext.js [hunk 3]
```diff
@@ -0,0 +1,5 @@
+import React from 'react';
+
+// Track which components mounted onto the DOM are coming from
+// Grommet (see FocusedContainer.js)
+export const GrommetContext = React.createContext(undefined);
```
Comment (ericsoderberghp) [inline src/js/components/Grommet/GrommetContext.js:5]: Wondering if "GrommetContext" is the right term here, perhaps "RootsContext"?

Comment (taysea) [reply src/js/components/Grommet/GrommetContext.js:5]: Good comment, I was trying to think what would be appropriate an appropriate name. I could see `Roots` working but it didn't click right off the bat. Let me brainstorm a little more while I address the other comments.

File: src/js/components/Form/__tests__/__snapshots__/Form-test-uncontrolled.js.snap (hunk 1 above)
Comment (taysea) [reply src/js/components/Form/__tests__/__snapshots__/Form-test-uncontrolled.js.snap:1944]: Good call, let me look into it.

File: src/js/components/Grommet/Grommet.js (hunk 2 above)
Comment (taysea) [reply src/js/components/Grommet/Grommet.js:102]: Hmm that's a good point. My understanding is that Context is contained, so I'm not sure if you can have an outer/inner context inherit the value of the other -- do you know?

Otherwise, we may need to rely more explicitly on a DOM attribute like "data-g-id" or something that we could scan the DOM for. What do you think?

Comment (ericsoderberghp) [reply src/js/components/Grommet/Grommet.js:102]: I think we have to avoid scanning the DOM. I was thinking that we just keep the additive nature, so a nested Grommet just adds another root to the array started by the outermost one.

File: src/js/components/Form/__tests__/__snapshots__/Form-test-uncontrolled.js.snap (hunk 1 above)
Comment (taysea) [reply src/js/components/Form/__tests__/__snapshots__/Form-test-uncontrolled.js.snap]: Resolved.

File: src/js/components/Grommet/GrommetContext.js (hunk 3 above)
Comment (taysea) [reply src/js/components/Grommet/GrommetContext.js]: I've gone with RootsContext. I think once I equated it to the concept of "root directory", it makes more sense. The other idea I had was `OriginContext`, but that didn't have any programming equivalent so it doesn't feel like it's a better option.

File: src/js/components/FocusedContainer.js [hunk 4]
```diff
@@ -18,44 +13,51 @@ export const FocusedContainer = ({
...
-      const child = ref.current;
-      getBodyChildElements()
-        .filter(isNotAncestorOf(child))
-        .forEach(makeNodeUnfocusable);
+  useEffect(() => {
+    if (
+      bodyOverflowStyle !== 'hidden' &&
+      !hidden &&
+      restrictScroll &&
+      trapFocus
```
Comment (ericsoderberghp) [inline src/js/components/FocusedContainer.js:29]: do we need `trapFocus` here?

File: src/js/components/FocusedContainer.js [hunk 5]
```diff
@@ -18,44 +13,51 @@ export const FocusedContainer = ({
...
+    ) {
+      setBodyOverflowStyle(document.body.style.overflow);
+      document.body.style.overflow = 'hidden';
+    }
 
-      if (restrictScroll && bodyOverflowStyle !== 'hidden') {
-        setBodyOverflowStyle(document.body.style.overflow);
-        document.body.style.overflow = 'hidden';
+    return () => {
+      if (restrictScroll) {
```
Comment (ericsoderberghp) [inline src/js/components/FocusedContainer.js:36]: It feels like this test should align more with the test at lines 26-28.

File: src/js/components/FocusedContainer.js [hunk 6]
```diff
@@ -18,44 +13,51 @@ export const FocusedContainer = ({
...
+        document.body.style.overflow = bodyOverflowStyle;
       }
     };
+  }, [bodyOverflowStyle, hidden, trapFocus, restrictScroll]);
 
+  useEffect(() => {
     const timer = setTimeout(() => {
-      if (!hidden && trapFocus) {
-        handleTrapFocus();
+      if (!hidden && trapFocus && root && root[0] !== null) {
```
Comment (ericsoderberghp) [inline src/js/components/FocusedContainer.js:44]: I would think `roots` would never have null entries.

File: src/js/components/FocusedContainer.js [hunk 7]
```diff
@@ -18,44 +13,51 @@ export const FocusedContainer = ({
...
+        root.forEach(makeNodeUnfocusable);
       }
     }, 0);
 
     return () => {
-      removeTrap();
+      // remove trap
+      if (root && root[0] !== null) makeNodeFocusable(root[root.length - 1]);
```
Comment (ericsoderberghp) [inline src/js/components/FocusedContainer.js:51]: I think a comment about why we only allow focus on the last root here would be helpful.

File: src/js/components/FocusedContainer.js (hunk 6 above)
Comment (taysea) [reply src/js/components/FocusedContainer.js:44]: In the browser, I wasn't having any issues, but when I ran the jest tests I was getting failures. Not sure if it was specific to the jest environment but adding a check resolved it

File: src/js/components/FocusedContainer.js [hunk 8]
```diff
@@ -18,44 +13,51 @@ export const FocusedContainer = ({
   const [bodyOverflowStyle, setBodyOverflowStyle] = useState('');
   const ref = useRef(null);
 
+  const root = useContext(RootsContext);
```
[the remark this answers is not in the document]
Comment (taysea) [reply src/js/components/FocusedContainer.js:16]: Good point, I'll update.

File: src/js/components/FocusedContainer.js (hunk 4 above)
Comment (taysea) [reply src/js/components/FocusedContainer.js:29]: I believe so, because if we're not trapping focus in the drop we would want the body to still be able to scroll in case an element on the page is receiving focus

File: src/js/components/FocusedContainer.js (hunk 5 above)
Comment (taysea) [reply src/js/components/FocusedContainer.js:36]: That makes sense. I've updated to condition.

File: src/js/components/FocusedContainer.js (hunk 4 above)
Comment (taysea) [reply src/js/components/FocusedContainer.js:29]: However, after thinking through it some more, it seems that "restrictScroll" should always be `true` when we want this to happen. The reason I had included it though was because it is part of the existing logic, but doing some testing I think we may be able to take it out.

File: src/js/components/FocusedContainer.js (hunk 7 above)
Comment (taysea) [reply src/js/components/FocusedContainer.js]: Done.

File: src/js/components/FocusedContainer.js (hunk 5 above)
Comment (taysea) [reply src/js/components/FocusedContainer.js]: Done.

File: src/js/components/FocusedContainer.js (hunk 8 above)
[the remark this answers is not in the document]
Comment (taysea) [reply src/js/components/FocusedContainer.js]: Done.

File: src/js/components/FocusedContainer.js [hunk 9]
```diff
@@ -18,44 +13,51 @@ export const FocusedContainer = ({
...
+  const [updatedRoot, setUpdatedRoot] = useState(root);
+
   useEffect(() => {
```
[the remark this answers is not in the document]
Comment (taysea) [reply src/js/components/FocusedContainer.js:18]: When I place it as a memo, it doesn't rerender to include the `ref.current`. Thoughts?

[the remark this answers is not in the document]
Comment (ericsoderberghp) [reply src/js/components/FocusedContainer.js:18]: I don't know enough to know why. I'm fine with leaving it as `useEffect`.

**Review by ericsoderberghp — APPROVED** [review]

**Review by halocline — COMMENTED** [review]
@taysea This is looking good. I like this approach. I made a first pass this afternoon, but would like to test manually tomorrow AM.

File: src/js/components/FocusedContainer.js [hunk 10]
```diff
@@ -18,44 +13,56 @@ export const FocusedContainer = ({
...
-      if (!hidden && trapFocus) {
-        handleTrapFocus();
+      if (!hidden && trapFocus && roots && roots[0] !== null) {
+        roots.forEach(makeNodeUnfocusable);
       }
     }, 0);
 
     return () => {
-      removeTrap();
+      // remove trap and restore ability to focus on the last root only
```
Comment (halocline) [inline src/js/components/FocusedContainer.js:54]: Question/Wondering: This is a style related comment, but since useEffects aren't named you don't get the benefit of having a clue of what's happening inside like you do with a well named function. 

The comment here is helpful, but I have been wondering about moving them just ahead of the useEffect so the reader knows why the effect is being run sooner.

In this particular instance, its positioning isn't really an issue, but when effects have more going on it can be quite helpful.

Just a wondering... / thought to share.

Comment (taysea) [reply src/js/components/FocusedContainer.js:54]: that's a good point -- in this case this comment is a little more specific to the clean-up function but I agree that a comment at the start of this effect could be helpful to say "toggle aria-hidden value for roots", and a similar pattern adopted in future

**Review by halocline — APPROVED** [review]
Looks good!
